### ISSUE TITLE
Item response theory model example/tutorial

### ISSUE THREAD

**Issue by dustinvtran** [tensorflow/probability#69]
See also https://github.com/blei-lab/edward/blob/master/examples/irt.py.

**Comment by joshchang** [comment]
I'm attempting to implement a graded response model in tfp as a learning problem but I cannot figure out where I am going wrong. Apologies in advance since I am new to tensorflow

The GRM has likelihood

```latex
\begin{align}
\Pr( X_{pi} = j \vert \theta_p, \tau_{i,j}, \tau_{i,j+1} , \lambda_i  )  &= \Pr(X_{pi}\geq j \vert\theta_p, \tau , \lambda )- \Pr(X_{pi}\geq j+1 \vert\theta_p, \tau , \lambda ) \nonumber\\
&=  S(\lambda_i(\theta_p-\tau_{i,j})) - S(\lambda_i(\theta_p-\tau_{i,j+1})),
\end{align}
```

where  S is the sigmoid function and I'm using the simple prior

```latex
\begin{align}
    \lambda_i&\sim C^+(0,5) \nonumber\\
    \tau_{i,0}\vert \mu_i & \sim N(\mu_i,3) \nonumber\\
    \mu_i &\sim N(0,3) \nonumber \\
    \tau_{i,j}|\tau_{i,j-1} &\sim N^+(\tau_{i,j-1},3)
\end{align}
```
I coded these up

```python

"""
Probabilities for a single person
discriminations   I (item)
difficulties I x K-1
abilities N 

returns
prob N x I x K
"""
def grm_model_prob(abilities, discriminations, difficulties, expanded=False):
    augmented_difficulties = tf.pad(difficulties, paddings=(
         (0, 0), (1, 0)), mode='constant', constant_values=-np.inf)
    augmented_difficulties = tf.pad(augmented_difficulties, paddings=(
         (0, 0), (0, 1)), mode='constant', constant_values=np.inf)
    offsets = augmented_difficulties[tf.newaxis,:,:] - abilities[:, tf.newaxis, tf.newaxis]
    scaled = offsets*discriminations[tf.newaxis, :, tf.newaxis]
    logits = 1.0/(1+tf.exp(scaled))
    probs = logits[:, :, :-1] - logits[:, :, 1:]
    
    return probs

"""
Probabilities for single items
discriminations   I (item)
difficulties  I x K - 1
abilities N

mu (difficulty local) I
tau (difficulty) I x K-2


"""
def joint_log_prob(responses, discriminations, difficulties, abilities, mu):
    return tf.reduce_sum(log_likelihood(responses, discriminations, difficulties, abilities)) + \
        joint_log_prior(discriminations, difficulties, abilities, mu)

def log_likelihood(responses, discriminations, difficulties, abilities):
    rv_responses = tfd.Categorical(grm_model_prob(
        abilities, discriminations, difficulties))
    return rv_responses.log_prob(responses)

def joint_log_prior(discriminations, difficulties, abilities, mu):
    rv_discriminations = tfd.Independent(tfd.HalfNormal(scale=tf.ones_like(discriminations)*5.), reinterpreted_batch_ndims=1)
    rv_difficulties0 =  tfd.MultivariateNormalDiag(loc=mu, scale_identity_multiplier=3.)
    rv_difficulties =  tfd.Independent(tfd.HalfNormal(scale=tf.ones_like(difficulties)[:,1:]), reinterpreted_batch_ndims=1)
    rv_abilities =  tfd.MultivariateNormalDiag(loc=tf.zeros_like(abilities), scale_identity_multiplier=3.)
    rv_mu =  tfd.MultivariateNormalDiag(loc=tf.zeros_like(mu),scale_identity_multiplier=3.)
    
    return tf.reduce_sum(rv_discriminations.log_prob(discriminations)) + \
        tf.reduce_sum(rv_difficulties0.log_prob(difficulties[:,0])) + \
        tf.reduce_sum(rv_difficulties.log_prob(difficulties[ :, 1:]-difficulties[ :, :-1])) + \
        tf.reduce_sum(rv_abilities.log_prob(abilities))
```

but I am unable to sample the posterior using either HMC or MH - all the proposals get rejected with the following samplers

```
hmc=tfp.mcmc.TransformedTransitionKernel(
    inner_kernel=tfp.mcmc.HamiltonianMonteCarlo(
        target_log_prob_fn=unnormalized_posterior_log_prob,
        num_leapfrog_steps=num_leapfrog_steps,
        step_size=0.5,
        state_gradients_are_stopped=True),
    bijector=unconstraining_bijectors)

mh = tfp.mcmc.TransformedTransitionKernel(
    inner_kernel=tfp.mcmc.RandomWalkMetropolis(
        target_log_prob_fn=unnormalized_posterior_log_prob),
    bijector=unconstraining_bijectors)
```

Any help is appreciated

**Comment by joshchang** [comment]
I'm trying to learn tfp and tensorflow by implementing the graded response irt model here: https://colab.research.google.com/drive/1ZAmkuQF2XoV0Jy9_EJtGTJrD3Tn4UlAJ

The model I'm trying to implement is described in that colab

 I'm unable to get the sampler working - either HMC or MH. HMC seems to give nan for the gradients and MH does give me something but it doesn't seem to be taking productive steps. The acceptance ratio also seems to be 0%. Am I doing something wrong? The model from the colab is here:

```python
"""
Probabilities for a single person
discriminations   I (item)
difficulties I x K-1
abilities N 

returns
prob N x I x K
"""
def grm_model_prob(abilities, discriminations, difficulties, expanded=False):
    offsets = difficulties[tf.newaxis,:,:] - abilities[:, tf.newaxis, tf.newaxis]
    scaled = offsets*discriminations[tf.newaxis, :, tf.newaxis]
    logits = 1.0/(1+tf.exp(scaled))
    logits = tf.pad(logits, paddings=(
        (0, 0), (0, 0), (1, 0)), mode='constant', constant_values=1.)
    logits = tf.pad(logits, paddings=(
        (0, 0), (0, 0), (0, 1)), mode='constant', constant_values=0.)
    probs = logits[:, :, :-1] - logits[:, :, 1:]
    
    return probs

"""
Probabilities for single items
discriminations   I (item)
difficulties  I x K - 1
abilities N

mu (difficulty local) I
tau (difficulty) I x K-2

"""
def joint_log_prob(responses, discriminations, difficulties0, ddifficulties, abilities, mu):
    # assemble the difficulties
    d0 = tf.concat([difficulties0[:,tf.newaxis],ddifficulties],axis=1)
    difficulties = tf.cumsum(d0,axis=1)
    return tf.reduce_sum(log_likelihood(responses, discriminations, difficulties, abilities)) + \
        joint_log_prior(discriminations, difficulties0, ddifficulties, abilities, mu)

def log_likelihood(responses, discriminations, difficulties, abilities):
    rv_responses = tfd.Categorical(grm_model_prob(
        abilities, discriminations, difficulties))
    return rv_responses.log_prob(responses)

def joint_log_prior(discriminations, difficulties0, ddifficulties, abilities, mu):
    rv_discriminations = tfd.HalfNormal(scale=tf.ones_like(discriminations))
    rv_difficulties0 =  tfd.Normal(loc=mu, scale=3.*tf.ones_like(difficulties0))
    rv_ddifficulties =  tfd.HalfNormal(scale=tf.ones_like(ddifficulties))
    rv_abilities =  tfd.Normal(loc=tf.zeros_like(abilities), scale=3.)
    rv_mu =  tfd.Normal(loc=tf.zeros_like(mu),scale=3.)
    
    return tf.reduce_sum(rv_discriminations.log_prob(discriminations)) + \
        tf.reduce_sum(rv_difficulties0.log_prob(difficulties0)) + \
        tf.reduce_sum(rv_ddifficulties.log_prob(ddifficulties)) + \
        tf.reduce_sum(rv_abilities.log_prob(abilities))

difficulties0 = np.sort(np.random.normal(size=(I, K-1)), axis=1)
abilities0 = np.random.normal(size=N)

#difficulties0 = np.zeros((I,K-1))
abilities0 = np.zeros(N)

# Set the chain's start state.
initial_chain_state = [
    tf.ones((I), name='init_discriminations'),
    tf.cast(difficulties0[:,0],tf.float32, name='init_difficulties0'), # may be causing problems
    tf.cast(difficulties0[:,1:]-difficulties0[:,:-1],tf.float32, name='init_ddifficulties'),
    tf.cast(abilities0,tf.float32, name='init_abilities'),
    tf.zeros((I),name='init_mu')
]

# Since MCMC operates over unconstrained space, we need to transform the
# samples so they live in real-space.
unconstraining_bijectors = [
    tfp.bijectors.Softplus(),       # R^+ \to R
    tfp.bijectors.Identity(),       # Maps R to R.
    tfp.bijectors.Softplus(),
    tfp.bijectors.Identity(),       # Maps R to R.
    tfp.bijectors.Identity()       # Maps R to R.
]

# Define a closure over our joint_log_prob.
unnormalized_posterior_log_prob = lambda *args: joint_log_prob(data, *args)

# Initialize the step_size. (It will be automatically adapted.)


number_of_steps=20000 
burnin=15000 
num_leapfrog_steps=5
hmc=tfp.mcmc.TransformedTransitionKernel(
    inner_kernel=tfp.mcmc.HamiltonianMonteCarlo(
        target_log_prob_fn=unnormalized_posterior_log_prob,
        num_leapfrog_steps=num_leapfrog_steps,
        step_size=0.1,
        state_gradients_are_stopped=True),
    bijector=unconstraining_bijectors)

mh = tfp.mcmc.TransformedTransitionKernel(
    inner_kernel=tfp.mcmc.RandomWalkMetropolis(
        target_log_prob_fn=unnormalized_posterior_log_prob),
    bijector=unconstraining_bijectors)

def trace_everything(states, previous_kernel_results):
  return previous_kernel_results

[discriminations,difficulties0,ddifficulties,abilities,mu], kernel_results = tfp.mcmc.sample_chain(
    num_results=number_of_steps,
    num_burnin_steps=burnin,
    current_state=initial_chain_state,
    trace_fn=trace_everything,
    kernel=mh)

with tf.device('/device:GPU:0'):
  [
    discriminations_,
    difficulties0_,
    ddifficulties_,
    abilities_,mu_,
    kernel_results_
  ] = evaluate([discriminations,difficulties0,ddifficulties,abilities,mu,kernel_results])
  
```

I appreciate any pointers
